src/huggingface_hub/hf_api.py

```diff
@@ -4772,6 +4772,92 @@ def delete_file(
+        this function have a trailing `/` (as in `folder/`),
+        this function will interpret it as `/*` (as in `folder/*`) to comply with Unix
+        patterns (Unix shell standards don't match `folder/a` to `folder/` but
+        match it to `folder/*`) when deleting all files in the folders(s).
+        By the previous corollary, the folder(s) will disappear because all
+        their files are deleted.
+
+        Args:
+            path_in_repo (`str`):
+                Relative folder path in the repo, for example: `"checkpoints/1fec34a"`.
+            repo_id (`str`):
+                The repository from which the folder will be deleted, for example:
+                `"username/custom_transformers"`
+            token (Union[bool, str, None], optional):
+                A valid user access token (string). Defaults to the locally saved
+                token, which is the recommended method for authentication (see
+                https://huggingface.co/docs/huggingface_hub/quick-start#authentication).
+                To disable authentication, pass `False`.
+                to the stored token.
+            repo_type (`str`, *optional*):
+                Set to `"dataset"` or `"space"` if the folder is in a dataset or
+                space, `None` or `"model"` if in a model. Default is `None`.
+            revision (`str`, *optional*):
+                The git revision to commit from. Defaults to the head of the `"main"` branch.
+            commit_message (`str`, *optional*):
+                The summary / title / first line of the generated commit. Defaults to
+                `f"Delete folder {path_in_repo} with huggingface_hub"`.
+            commit_description (`str` *optional*)
+                The description of the generated commit.
+            create_pr (`boolean`, *optional*):
+                Whether or not to create a Pull Request with that commit. Defaults to `False`.
+                If `revision` is not set, PR is opened against the `"main"` branch. If
+                `revision` is set and is a branch, PR is opened against this branch. If
+                `revision` is set and is not a branch name (example: a commit oid), an
+                `RevisionNotFoundError` is returned by the server.
+            parent_commit (`str`, *optional*):
+                The OID / SHA of the parent commit, as a hexadecimal string. Shorthands (7 first characters) are also supported.
+                If specified and `create_pr` is `False`, the commit will fail if `revision` does not point to `parent_commit`.
+                If specified and `create_pr` is `True`, the pull request will be created from `parent_commit`.
+                Specifying `parent_commit` ensures the repo has not changed before committing the changes, and can be
+                especially useful if the repo is updated / committed to concurrently.
+        """
+        patterns = [p + "*" if p[-1] == "/" else p for p in patterns]
```

Comment:
Are we sure this is needed? When creating a commit, a `CommitOperationDelete` operation with `"folder/"` (no `*`) should already delete the entire folder for us. So I'm surprised if `"*"` is needed. If that's the case, it should be fixed in `_prepare_folder_deletions` directly.

Reply:
Yep it is needed, as I documented it in the docstring and in the PR description : 

Here is what you get if you comment this line : 
![image](https://github.com/huggingface/huggingface_hub/assets/76554137/6f1d30e3-d138-4205-af2f-84d2b5afdb8b)

Fixing it in `_prepare_folder_deletions` will alter the `upload` command behavior related to deletion (see PR description). Should we still do it ?

Reply:
Indeed. I had a look into it and I think it should be done in `filter_repo_objects` directly. I just pushed https://github.com/huggingface/huggingface_hub/pull/2280/commits/19d6b199473e5f11ec563fa90510a6c731df6363 to your PR, hope you don't mind.